onnxruntime/core/providers/cpu/math/gemm.cc

```diff
@@ -138,8 +166,69 @@
                 thread_pool);
 }
 
+template <>
+void Gemm<MLFloat16>::ComputeGemm(CBLAS_TRANSPOSE trans_a, CBLAS_TRANSPOSE trans_b,
+                                  ptrdiff_t M, ptrdiff_t N, ptrdiff_t K,
+                                  MLFloat16 alpha,
+                                  const MLFloat16* a_data, const MLFloat16* b_data,
+                                  MLFloat16 beta,
+                                  const MLFloat16* c_data, const TensorShape* c_shape,
+                                  MLFloat16* y_data,
+                                  concurrency::ThreadPool* thread_pool) {
+  // if input is empty tensor, return directly as nothing need to be calculated.
+  if (M == 0 || N == 0)
+    return;
+
+#if defined(__GNUC__) && defined(HAS_CLASS_MEMACCESS)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wclass-memaccess"
+#endif
+  // MLFloat16's constructor is explict, so here we need to use memset
+  if (c_data == nullptr)
+    memset(&beta, 0, sizeof(MLFloat16));
+#if defined(__GNUC__) && defined(HAS_CLASS_MEMACCESS)
+#pragma GCC diagnostic pop
+#endif
+#ifdef MLAS_F16VEC_INTRINSICS_SUPPORTED
```

Comment:
Why do you need all these #ifdef? As we discussed before, in cpu_execution_provider.cc, all fp16 gemm operator declaration is already guarded by compilation flat, and the register function RegisterFp16Kernels are not only guarded by this flag, but also by dynamic capability check MlasFp16AccelerationSupported.

So this function is simply not reachable if fp16 vector support is not present. And all these structural effort is to make the operator implementations simpler. Why do you need to bring the complexity back here?

Reply:
Maybe there was some misunderstanding. I thought you told me that in some cases we still need to fallback to eigen? In corner cases that this kernel is registered(because the hardware supports fp16) but mlas hasn't implemented these corner cases, we may still want to fall back to eigen.